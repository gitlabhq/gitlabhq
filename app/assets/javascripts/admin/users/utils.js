import { queryToObject } from '~/lib/utils/url_utility';
import getSoloOwnedOrganizationsQuery from '~/admin/users/graphql/queries/get_solo_owned_organizations.query.graphql';
import { convertToGraphQLId } from '~/graphql_shared/utils';
import { TYPENAME_USER } from '~/graphql_shared/constants';
import {
  TOKENS,
  SOLO_OWNED_ORGANIZATIONS_EMPTY,
  SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
} from 'ee_else_ce/admin/users/constants';

export const generateUserPaths = (paths, id) => {
  return Object.fromEntries(
    Object.entries(paths).map(([action, genericPath]) => {
      return [action, genericPath.replace('id', id)];
    }),
  );
};

/**
 * @typedef {{type: string, value: {data: string, operator: string}}} Token
 */

/**
 * Initialize token values based on the URL parameters
 * @param {string} query - document.location.searchd
 *
 * @returns {{tokens: Array<string|Token>, sort: string}}
 */
export function initializeValuesFromQuery(query = document.location.search) {
  const tokens = [];

  const { filter, search_query: searchQuery, sort } = queryToObject(query);

  if (filter) {
    const token = TOKENS.find(({ options }) => options.some(({ value }) => value === filter));

    if (token) {
      tokens.push({
        type: token.type,
        value: {
          data: filter,
          operator: token.operators[0].value,
        },
      });
    }
  }

  if (searchQuery) {
    tokens.push(searchQuery);
  }

  return { tokens, sort };
}

export const getSoloOwnedOrganizations = async (apolloClient, userId) => {
  if (!window.gon?.features?.uiForOrganizations) {
    return Promise.resolve(SOLO_OWNED_ORGANIZATIONS_EMPTY);
  }

  const {
    data: {
      user: { organizations },
    },
  } = await apolloClient.query({
    query: getSoloOwnedOrganizationsQuery,
    variables: {
      id: convertToGraphQLId(TYPENAME_USER, userId),
      first: SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
    },
  });

  return organizations;
};
