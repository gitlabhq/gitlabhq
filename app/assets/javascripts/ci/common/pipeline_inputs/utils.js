import { isEqual } from 'lodash';

/**
 * Generates skeleton rect props for the skeleton loader based on column and row indices
 *
 * @param {Number} columnIndex - The column index (0-based)
 * @param {Number} rowIndex - The row index (0-based)
 * @returns {Object} - The props for the skeleton rect
 */
export const getSkeletonRectProps = (columnIndex, rowIndex) => {
  return {
    x: `${columnIndex * 25.5}%`,
    y: rowIndex * 10,
    width: '23%',
    height: 6,
    rx: 2,
    ry: 2,
  };
};

/**
 * Formats a value for display in the input preview
 * @param {*} value - The value to format
 * @returns {string} - Formatted string representation
 */
export function formatValue(value) {
  if (value === null) return 'null';
  if (value === undefined) return 'undefined';
  if (typeof value === 'string') return `"${value}"`;
  if (typeof value === 'object') return JSON.stringify(value);
  return String(value);
}

/**
 * Checks if an input value has changed from its default
 * @param {*} value - Current value
 * @param {*} defaultValue - Default value
 * @returns {boolean} - True if values are different
 */
export function hasValueChanged(value, defaultValue) {
  if (typeof value === 'object' || typeof defaultValue === 'object') {
    return !isEqual(value, defaultValue);
  }
  return value !== defaultValue;
}

/**
 * Formats value lines for diff display
 * @param {Object} input - Input object with value, default, etc.
 * @param {boolean} isChanged - Whether the value has changed
 * @returns {Array} - Array of line objects
 */
export function formatValueLines(input, isChanged) {
  const lines = [];
  const formattedValue = formatValue(input.value);
  const formattedDefault = formatValue(input.default);

  if (isChanged) {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `-   value: ${formattedDefault}`,
      type: 'old',
    });
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `+   value: ${formattedValue}`,
      type: 'new',
    });
  } else {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `    value: ${formattedValue}`,
      type: '',
    });
  }

  return lines;
}

/**
 * Formats metadata lines (type, description)
 * @param {Object} input - Input object
 * @returns {Array} - Array of line objects
 */
export function formatMetadataLines(input) {
  const lines = [];

  if (input.type) {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `    type: "${input.type}"`,
      type: '',
    });
  }

  if (input.description) {
    lines.push({
      // eslint-disable-next-line @gitlab/require-i18n-strings
      content: `    description: "${input.description}"`,
      type: '',
    });
  }

  return lines;
}

/**
 * Formats a single input into display lines
 * @param {Object} input - Input object
 * @returns {Array} - Array of line objects
 */
export function formatInputLines(input) {
  const lines = [];
  const isChanged = hasValueChanged(input.value, input.default);

  lines.push({
    content: `${input.name}:`,
    type: '',
  });

  lines.push(...formatValueLines(input, isChanged));

  lines.push(...formatMetadataLines(input));

  lines.push({
    content: '',
    type: '',
  });

  return lines;
}

/**
 * Formats all inputs into display lines
 * @param {Array} inputs - Array of input objects
 * @returns {Array} - Array of line objects for display
 */
export function formatInputsForDisplay(inputs) {
  return inputs.flatMap(formatInputLines);
}
