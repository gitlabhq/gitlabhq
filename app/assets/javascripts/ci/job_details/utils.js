export const compactJobLog = (jobLog) => {
  const compactedLog = [];

  jobLog.forEach((obj) => {
    // push header section line
    if (obj.line && obj.isHeader) {
      compactedLog.push(obj.line);
    }

    // push lines within section header
    if (obj.lines?.length > 0) {
      compactedLog.push(...obj.lines);
    }

    // push lines from plain log
    if (!obj.lines && obj.content.length > 0) {
      compactedLog.push(obj);
    }
  });

  return compactedLog;
};

export const filterAnnotations = (annotations, type) => {
  return [...annotations]
    .sort((a, b) => a.name.localeCompare(b.name))
    .flatMap((annotationList) => annotationList.data)
    .flatMap((annotation) => annotation[type] ?? []);
};

export const getLineText = (line) => {
  return line.content.map((content) => content.text).join('');
};
