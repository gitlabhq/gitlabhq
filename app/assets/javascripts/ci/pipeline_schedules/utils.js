export const updateScheduleNodes = (schedules = [], updatedSchedule = {}) => {
  return schedules.map((schedule) => {
    if (schedule.id === updatedSchedule.id) {
      return {
        ...schedule,
        lastPipeline: updatedSchedule.lastPipeline,
      };
    }

    return schedule;
  });
};
