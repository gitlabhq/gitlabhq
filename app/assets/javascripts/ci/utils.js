import * as Sentry from '~/sentry/sentry_browser_wrapper';

export const reportToSentry = (component, error) => {
  Sentry.captureException(error, {
    tags: {
      component,
    },
  });
};
