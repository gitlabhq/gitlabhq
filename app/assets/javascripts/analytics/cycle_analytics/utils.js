import { parseSeconds, newDate } from '~/lib/utils/datetime_utility';
import { formatTimeAsSummary } from '~/lib/utils/datetime/date_format_utility';
import { joinPaths } from '~/lib/utils/url_utility';

/**
 * Takes the stages and median data, combined with the selected stage, to build an
 * array which is formatted to proivde the data required for the path navigation.
 *
 * @param {Array} stages - The stages available to the group / project
 * @param {Object} medians - The median values for the stages available to the group / project
 * @param {Object} stageCounts - The total item count for the stages available
 * @param {Object} selectedStage - The currently selected stage
 * @returns {Array} An array of stages formatted with data required for the path navigation
 */
export const transformStagesForPathNavigation = ({
  stages,
  medians,
  stageCounts = {},
  selectedStage,
}) => {
  const formattedStages = stages.map((stage) => {
    return {
      metric: medians[stage?.id],
      selected: stage?.id && stage?.id === selectedStage?.id,
      stageCount: stageCounts && stageCounts[stage?.id],
      icon: null,
      ...stage,
    };
  });

  return formattedStages;
};

/**
 * Takes a raw median value in seconds and converts it to a string representation
 * ie. converts 172800 => 2d (2 days)
 *
 * @param {Number} Median - The number of seconds for the median calculation
 * @returns {String} String representation ie 2w
 */
export const medianTimeToParsedSeconds = (value) =>
  formatTimeAsSummary({
    ...parseSeconds(value, { daysPerWeek: 7, hoursPerDay: 24 }),
    seconds: value,
  });

/**
 * Takes the raw median value arrays and converts them into a useful object
 * containing the string for display in the path navigation
 * ie. converts [{ id: 'test', value: 172800 }] => { 'test': '2d' }
 *
 * @param {Array} Medians - Array of stage median objects, each contains a `id`, `value` and `error`
 * @returns {Object} Returns key value pair with the stage name and its display median value
 */
export const formatMedianValues = (medians = []) =>
  medians.reduce((acc, { id, value = 0 }) => {
    return {
      ...acc,
      [id]: value ? medianTimeToParsedSeconds(value) : '-',
    };
  }, {});

export const filterStagesByHiddenStatus = (stages = [], isHidden = true) =>
  stages.filter(({ hidden = false }) => hidden === isHidden);

/**
 * Builds the initial data object for Value Stream Analytics with data loaded from the backend
 *
 * @param {Object} dataset - dataset object paseed to the frontend via data-* properties
 * @returns {Object} - The initial data to load the app with
 */
export const buildCycleAnalyticsInitialData = ({
  projectId,
  groupPath,
  stage,
  createdAfter,
  createdBefore,
  namespaceName,
  namespacePath,
  namespaceRestApiRequestPath,
} = {}) => {
  return {
    projectId: parseInt(projectId, 10),
    groupPath,
    namespace: {
      name: namespaceName,
      path: namespacePath,
      restApiRequestPath: namespaceRestApiRequestPath,
    },
    createdAfter: newDate(createdAfter),
    createdBefore: newDate(createdBefore),
    selectedStage: stage ? JSON.parse(stage) : null,
  };
};

export const constructPathWithNamespace = ({ restApiRequestPath }, endpoint) =>
  joinPaths('/', restApiRequestPath, endpoint);
