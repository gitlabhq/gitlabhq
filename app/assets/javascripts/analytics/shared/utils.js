import dateFormat from '~/lib/dateformat';
import { SECONDS_IN_DAY } from '~/lib/utils/datetime_utility';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import { joinPaths } from '~/lib/utils/url_utility';
import { urlQueryToFilter } from '~/vue_shared/components/filtered_search_bar/filtered_search_utils';
import {
  dateFormats,
  FLOW_METRICS,
  MAX_METRIC_PRECISION,
  UNITS,
  VALUE_STREAM_METRIC_DISPLAY_UNITS,
  VALUE_STREAM_METRIC_METADATA,
  VALUE_STREAM_METRIC_TILE_METADATA,
} from './constants';

export const filterBySearchTerm = (data = [], searchTerm = '', filterByKey = 'name') => {
  if (!searchTerm?.length) return data;
  return data.filter((item) => item[filterByKey].toLowerCase().includes(searchTerm.toLowerCase()));
};

export const toYmd = (date) => dateFormat(date, dateFormats.isoDate);

/**
 * Takes a url and extracts query parameters used for the shared
 * filter bar
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractFilterQueryParameters = (url = '') => {
  const {
    source_branch_name: selectedSourceBranch = null,
    target_branch_name: selectedTargetBranch = null,
    author_username: selectedAuthor = null,
    milestone_title: selectedMilestone = null,
    assignee_username: selectedAssigneeList = [],
    label_name: selectedLabelList = [],
  } = urlQueryToFilter(url);

  return {
    selectedSourceBranch,
    selectedTargetBranch,
    selectedAuthor,
    selectedMilestone,
    selectedAssigneeList,
    selectedLabelList,
  };
};

/**
 * Takes a url and extracts sorting and pagination query parameters into an object
 *
 * @param {string} url The URL to extract query parameters from
 * @returns {Object}
 */
export const extractPaginationQueryParameters = (url = '') => {
  const { sort, direction, page } = urlQueryToFilter(url);
  return {
    sort: sort?.value || null,
    direction: direction?.value || null,
    page: page?.value || null,
  };
};

export const getDataZoomOption = ({
  totalItems = 0,
  maxItemsPerPage = 40,
  dataZoom = [{ type: 'slider', bottom: 10, start: 0 }],
}) => {
  if (totalItems <= maxItemsPerPage) {
    return {};
  }

  const intervalEnd = Math.ceil((maxItemsPerPage / totalItems) * 100);

  return dataZoom.map((item) => {
    return {
      ...item,
      end: intervalEnd,
    };
  });
};

/**
 * Formats any valid number as percentage
 *
 * @param {number|string} decimalValue Decimal value between 0 and 1 to be converted to a percentage
 * @param {number} precision The number of decimal places to round to
 *
 * @returns {string} Returns a formatted string multiplied by 100
 */
export const formatAsPercentageWithoutSymbol = (decimalValue = 0, precision = 1) => {
  const parsed = Number.isNaN(Number(decimalValue)) ? 0 : decimalValue;
  return (parsed * 100).toFixed(precision);
};

/**
 * Converts a time in seconds to number of days, with variable precision
 *
 * @param {Number} seconds Time in seconds
 * @param {Number} precision Specifies the number of digits after the decimal
 *
 * @returns {String} The number of days
 */
export const secondsToDays = (seconds, precision = 1) =>
  (seconds / SECONDS_IN_DAY).toFixed(precision);

export const scaledValueForDisplay = (value, units, precision = MAX_METRIC_PRECISION) => {
  switch (units) {
    case UNITS.PERCENT:
      return formatAsPercentageWithoutSymbol(value);
    case UNITS.DAYS:
      return secondsToDays(value, precision);
    default:
      return value;
  }
};

const prepareMetricValue = ({ identifier, value, unit }) => {
  // NOTE: the flow metrics graphql endpoint returns values already scaled for display
  if (!value) {
    // ensures we return `-` for 0/null etc
    return '-';
  }
  return Object.values(FLOW_METRICS).includes(identifier)
    ? value
    : scaledValueForDisplay(value, unit);
};

/**
 * Prepares metric data to be rendered in the metric_tile component
 *
 * @param {MetricData[]} data - The metric data to be rendered
 * @returns {TransformedMetricData[]} An array of metrics ready to render in the metric_tile
 */
export const rawMetricToMetricTile = (metric) => {
  const { identifier, value, ...metricRest } = metric;
  const { unit, label, ...metadataRest } = VALUE_STREAM_METRIC_TILE_METADATA[identifier];
  return {
    ...metadataRest,
    ...metricRest,
    title: label,
    identifier,
    label,
    unit: VALUE_STREAM_METRIC_DISPLAY_UNITS[unit],
    value: prepareMetricValue({ value, unit, identifier }),
  };
};

/**
 * Generates a URL link to the VSD dashboard based on the
 * namespace path passed into the method.
 *
 * @param {String} namespacePath - Path of the specified namespace
 * @param {Boolean} isProjectNamespace
 * @returns a URL or blank string if there is no namespacePath set
 */
export const generateValueStreamsDashboardLink = (
  namespacePath = null,
  isProjectNamespace = false,
) => {
  if (!namespacePath) return '';

  const dashboardsSlug = '/-/analytics/dashboards/value_streams_dashboard';
  const formattedNamespacePath = isProjectNamespace ? namespacePath : `groups/${namespacePath}`;
  const segments = [gon.relative_url_root || '', '/', formattedNamespacePath, dashboardsSlug];

  return joinPaths(...segments);
};

/**
 * Generates a URL link to the metric's report based on the namespace type.
 * @param {String} metricId â€“ Metric's identifier
 * @param {String} namespacePath - Path of the specified namespace
 * @param {Boolean} isProjectNamespace
 * @returns {String} URL link to the metric's report
 */
export const generateMetricLink = ({
  metricId = null,
  namespacePath = null,
  isProjectNamespace = false,
} = {}) => {
  const { groupLink, projectLink } = VALUE_STREAM_METRIC_METADATA[metricId] ?? {};
  const hasLinks = groupLink || projectLink;

  if (!hasLinks || !namespacePath) return '';

  return joinPaths(
    '/',
    gon.relative_url_root,
    !isProjectNamespace ? 'groups' : '',
    namespacePath,
    isProjectNamespace ? projectLink : groupLink,
  );
};

/**
 * Extracts the relevant feature and license flags needed for VSA
 *
 * @param {Object} gon the global `window.gon` object populated when the page loads
 * @returns an object containing the extracted feature flags and their boolean status
 */
export const extractVSAFeaturesFromGON = () => ({
  // licensed feature toggles
  cycleAnalyticsForGroups: Boolean(gon?.licensed_features?.cycleAnalyticsForGroups),
  cycleAnalyticsForProjects: Boolean(gon?.licensed_features?.cycleAnalyticsForProjects),
  groupLevelAnalyticsDashboard: Boolean(gon?.licensed_features?.groupLevelAnalyticsDashboard),
});

/**
 * Takes a raw GraphQL response which could contain data for a group or project namespace,
 * and returns the data for the namespace which is present in the response.
 *
 * @param {Object} params
 * @param {string} params.resultKey - The data to be extracted from the namespace.
 * @param {Object} params.result
 * @param {Object} params.result.data
 * @param {Object} params.result.data.group - The group GraphQL response.
 * @param {Object} params.result.data.project - The project GraphQL response.
 * @returns {Object} The data extracted from either group[resultKey] or project[resultKey].
 */
export const extractQueryResponseFromNamespace = ({ result, resultKey }) => {
  const { group = null, project = null } = result.data;
  if (group || project) {
    const namespace = group ?? project;
    return namespace[resultKey] || {};
  }
  return {};
};

/**
 * Takes the raw snake_case query parameters and extracts + converts the relevant values
 * for the overview metrics component
 * @param {Object} params - Object containing the supported query parameters
 * @param {Date} params.created_before
 * @param {Date} params.created_after
 * @param {string} params.author_username
 * @param {string} params.milestone_title
 * @param {Array} params.label_name
 * @param {Array} params.assignee_username
 *
 * @returns {Object} CamelCased parameter names
 */
export const overviewMetricsRequestParams = (params = {}) => {
  const {
    createdAfter: startDate,
    createdBefore: endDate,
    labelName: labelNames,
    assigneeUsername: assigneeUsernames,
    ...rest
  } = convertObjectPropsToCamelCase(params);
  return {
    startDate,
    endDate,
    labelNames,
    assigneeUsernames,
    ...rest,
  };
};
