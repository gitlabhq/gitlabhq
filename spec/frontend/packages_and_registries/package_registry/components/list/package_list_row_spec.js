import { GlFormCheckbox, GlSprintf, GlTruncate, GlBadge } from '@gitlab/ui';
import Vue from 'vue';
import VueRouter from 'vue-router';
import { RouterLinkStub } from '@vue/test-utils';
import ProtectedBadge from '~/vue_shared/components/badges/protected_badge.vue';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import { getIdFromGraphQLId } from '~/graphql_shared/utils';
import PackagesListRow from '~/packages_and_registries/package_registry/components/list/package_list_row.vue';
import PackageTags from '~/packages_and_registries/shared/components/package_tags.vue';
import PublishMessage from '~/packages_and_registries/shared/components/publish_message.vue';
import PublishMethod from '~/packages_and_registries/package_registry/components/list/publish_method.vue';
import {
  PACKAGE_ERROR_STATUS,
  PACKAGE_DEPRECATED_STATUS,
} from '~/packages_and_registries/package_registry/constants';

import ListItem from '~/vue_shared/components/registry/list_item.vue';
import {
  conanMetadata,
  linksData,
  packageData,
  packagePipelines,
  packageProject,
  packageTags,
} from '../../mock_data';

Vue.use(VueRouter);

describe('packages_list_row', () => {
  let wrapper;

  const defaultProvide = {
    isGroupPage: false,
    canDeletePackages: true,
  };

  const packageWithoutTags = { ...packageData(), project: packageProject(), ...linksData };
  const packageWithTags = { ...packageWithoutTags, tags: { nodes: packageTags() } };

  const findPackageTags = () => wrapper.findComponent(PackageTags);
  const findDeprecatedBadge = () => wrapper.findComponent(GlBadge);
  const findDeleteDropdown = () => wrapper.findByTestId('delete-dropdown');
  const findDeleteButton = () => wrapper.findByTestId('action-delete');
  const findErrorMessage = () => wrapper.findByTestId('error-message');
  const findPackageType = () => wrapper.findByTestId('package-type');
  const findPackageLink = () => wrapper.findByTestId('details-link');
  const findWarningIcon = () => wrapper.findByTestId('warning-icon');
  const findLeftSecondaryInfos = () => wrapper.findByTestId('left-secondary-infos');
  const findPackageVersion = () => findLeftSecondaryInfos().findComponent(GlTruncate);
  const findPublishMessage = () => wrapper.findComponent(PublishMessage);
  const findPublishMethod = () => wrapper.findComponent(PublishMethod);
  const findListItem = () => wrapper.findComponent(ListItem);
  const findBulkDeleteAction = () => wrapper.findComponent(GlFormCheckbox);
  const findPackageName = () => wrapper.findByTestId('package-name');
  const findConanMetadata = () => wrapper.findByTestId('conan-metadata');
  const findPackageUsername = () => wrapper.findByTestId('package-username');
  const findPackageChannel = () => wrapper.findByTestId('package-channel');

  const mountComponent = ({
    packageEntity = packageWithoutTags,
    selected = false,
    provide = defaultProvide,
  } = {}) => {
    wrapper = shallowMountExtended(PackagesListRow, {
      provide,
      stubs: {
        ListItem,
        GlSprintf,
        RouterLink: RouterLinkStub,
        GlBadge,
      },
      propsData: {
        packageEntity,
        selected,
      },
    });
  };

  it('renders', () => {
    mountComponent();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('has a link to navigate to the details page', () => {
    mountComponent();

    expect(findPackageLink().props()).toMatchObject({
      to: { name: 'details', params: { id: getIdFromGraphQLId(packageWithoutTags.id) } },
    });
  });

  it('lists the package name', () => {
    mountComponent();

    expect(findPackageName().text()).toBe('@gitlab-org/package-15');
  });

  describe('tags', () => {
    it('renders package tags when a package has tags', () => {
      mountComponent({ packageEntity: packageWithTags });

      expect(findPackageTags().exists()).toBe(true);
    });

    it('does not render when there are no tags', () => {
      mountComponent();

      expect(findPackageTags().exists()).toBe(false);
    });
  });

  describe('delete dropdown', () => {
    it('does not exist when package cannot be destroyed', () => {
      mountComponent({
        packageEntity: { ...packageWithoutTags, userPermissions: { destroyPackage: false } },
      });

      expect(findDeleteDropdown().exists()).toBe(false);
    });

    it('exists when package can be destroyed', () => {
      mountComponent();

      expect(findDeleteDropdown().props()).toMatchObject({
        category: 'tertiary',
        icon: 'ellipsis_v',
        textSrOnly: true,
        noCaret: true,
        toggleText: 'More actions',
      });
    });
  });

  describe('delete button', () => {
    it('does not exist when package cannot be destroyed', () => {
      mountComponent({
        packageEntity: { ...packageWithoutTags, userPermissions: { destroyPackage: false } },
      });

      expect(findDeleteButton().exists()).toBe(false);
    });

    it('exists and has the correct text', () => {
      mountComponent({ packageEntity: packageWithoutTags });

      expect(findDeleteButton().exists()).toBe(true);
      expect(findDeleteButton().text()).toBe('Delete package');
    });

    it('emits the delete event when the delete button is clicked', () => {
      mountComponent({ packageEntity: packageWithoutTags });

      findDeleteButton().vm.$emit('action');

      expect(wrapper.emitted('delete')).toHaveLength(1);
    });
  });

  describe(`when the package is in ${PACKAGE_ERROR_STATUS} status`, () => {
    beforeEach(() => {
      mountComponent({
        packageEntity: {
          ...packageWithoutTags,
          status: PACKAGE_ERROR_STATUS,
          _links: {
            webPath: null,
          },
        },
      });
    });

    it('lists the package name', () => {
      expect(findPackageName().text()).toBe('@gitlab-org/package-15');
    });

    it('does not show the publish method', () => {
      expect(findPublishMethod().exists()).toBe(false);
    });

    it('does not show published message', () => {
      expect(findPublishMessage().exists()).toBe(false);
    });

    it('does not have a link to navigate to the details page', () => {
      expect(findPackageLink().exists()).toBe(false);
    });

    it('has a warning icon', () => {
      const icon = findWarningIcon();
      expect(icon.props('name')).toBe('warning');
    });

    it('renders error message text', () => {
      expect(findErrorMessage().text()).toEqual(
        'Error publishing · Invalid Package: failed metadata extraction',
      );
    });

    describe('with custom error message', () => {
      it('renders error message text', () => {
        mountComponent({
          packageEntity: {
            ...packageWithoutTags,
            status: PACKAGE_ERROR_STATUS,
            statusMessage: 'custom error message',
            _links: {
              webPath: null,
            },
          },
        });

        expect(findErrorMessage().text()).toEqual('Error publishing · custom error message');
      });
    });

    it('has a delete dropdown', () => {
      expect(findDeleteDropdown().exists()).toBe(true);
    });
  });

  describe('left action template', () => {
    it('does not render checkbox if not permitted', () => {
      mountComponent({
        provide: {
          ...defaultProvide,
          canDeletePackages: false,
        },
      });

      expect(findBulkDeleteAction().exists()).toBe(false);
    });

    it('renders checkbox', () => {
      mountComponent();

      expect(findBulkDeleteAction().exists()).toBe(true);
      expect(findBulkDeleteAction().attributes('checked')).toBeUndefined();
    });

    it('emits select when checked', () => {
      mountComponent();

      findBulkDeleteAction().vm.$emit('change');

      expect(wrapper.emitted('select')).toHaveLength(1);
    });

    it('renders checkbox in selected state if selected', () => {
      mountComponent({
        selected: true,
      });

      expect(findBulkDeleteAction().attributes('checked')).toBe('true');
      expect(findListItem().props()).toMatchObject({
        selected: true,
      });
    });
  });

  describe('secondary left info', () => {
    it('has the package version', () => {
      mountComponent();

      expect(findPackageVersion().props()).toMatchObject({
        text: packageWithoutTags.version,
        withTooltip: true,
      });
    });

    it('has package type with middot', () => {
      mountComponent();

      expect(findPackageType().text()).toBe(`· ${packageWithoutTags.packageType.toLowerCase()}`);
    });
  });

  describe('right info', () => {
    const projectPageProps = {
      projectName: '',
      projectUrl: '',
      publishDate: packageWithoutTags.createdAt,
    };

    it('has publish method component', () => {
      mountComponent({
        packageEntity: { ...packageWithoutTags, pipelines: { nodes: packagePipelines() } },
      });

      expect(findPublishMethod().props('pipeline')).toEqual(packagePipelines()[0]);
    });

    it('if the package is published through CI sets author on PublishMessage component', () => {
      mountComponent({
        packageEntity: { ...packageWithoutTags, pipelines: { nodes: packagePipelines() } },
      });

      expect(findPublishMessage().props()).toStrictEqual({
        author: 'Administrator',
        ...projectPageProps,
      });
    });

    it('if the package is published manually then does not set author on PublishMessage component', () => {
      mountComponent({
        packageEntity: { ...packageWithoutTags },
      });

      expect(findPublishMessage().props()).toStrictEqual({
        author: '',
        ...projectPageProps,
      });
    });

    describe('PublishMessage component for group page', () => {
      const groupPageProps = {
        projectName: packageWithoutTags.project.name,
        projectUrl: packageWithoutTags.project.webUrl,
        publishDate: packageWithoutTags.createdAt,
      };

      it('if the package is published through CI sets project name, url and author', () => {
        mountComponent({
          provide: {
            ...defaultProvide,
            isGroupPage: true,
          },
          packageEntity: { ...packageWithoutTags, pipelines: { nodes: packagePipelines() } },
        });

        expect(findPublishMessage().props()).toStrictEqual({
          author: 'Administrator',
          ...groupPageProps,
        });
      });

      it('if the package is published manually passes show project name, url and does not set author', () => {
        mountComponent({
          provide: {
            ...defaultProvide,
            isGroupPage: true,
          },
          packageEntity: { ...packageWithoutTags },
        });

        expect(findPublishMessage().props()).toStrictEqual({
          author: '',
          ...groupPageProps,
        });
      });
    });
  });

  describe('badge "protected"', () => {
    const mountComponentForBadgeProtected = ({ packageEntityProtectionRuleExists = true } = {}) =>
      mountComponent({
        packageEntity: {
          ...packageWithoutTags,
          protectionRuleExists: packageEntityProtectionRuleExists,
        },
        provide: {
          ...defaultProvide,
        },
      });

    const findProtectedBadge = () => wrapper.findComponent(ProtectedBadge);

    describe('when package is protected', () => {
      it('shows badge', () => {
        mountComponentForBadgeProtected();

        expect(findProtectedBadge().exists()).toBe(true);
        expect(findProtectedBadge().props('tooltipText')).toMatch(
          'A protection rule exists for this package.',
        );
      });
    });

    describe('when package is not protected', () => {
      it('does not show badge', () => {
        mountComponentForBadgeProtected({ packageEntityProtectionRuleExists: false });

        expect(findProtectedBadge().exists()).toBe(false);
      });
    });
  });

  describe('deprecated badge', () => {
    it('is not rendered by default', () => {
      mountComponent();

      expect(findDeprecatedBadge().exists()).toBe(false);
    });

    describe('when package has deprecated status', () => {
      beforeEach(() => {
        mountComponent({
          packageEntity: {
            ...packageWithoutTags,
            status: PACKAGE_DEPRECATED_STATUS,
          },
        });
      });

      it('renders GlBadge component', () => {
        expect(findDeprecatedBadge().props('variant')).toBe('warning');
      });

      it('renders the text `deprecated`', () => {
        expect(findDeprecatedBadge().text()).toBe('deprecated');
      });
    });
  });

  describe('conan metadata', () => {
    it('does not render `packageUsername` and `packageChannel` for any packages other than conan', () => {
      mountComponent();

      expect(findConanMetadata().exists()).toBe(false);
    });

    it('renders correct `packageUsername` and `packageChannel` when package type is conan', () => {
      mountComponent({
        packageEntity: {
          ...packageWithoutTags,
          packageType: 'CONAN',
          metadata: conanMetadata(),
        },
      });

      expect(findConanMetadata().exists()).toBe(true);
      expect(findPackageUsername().text()).toBe('gitlab-org+gitlab-test');
      expect(findPackageChannel().text()).toBe('stable');
    });

    it('hides Conan metadata with default values', () => {
      mountComponent({
        packageEntity: {
          ...packageWithoutTags,
          packageType: 'CONAN',
          metadata: { ...conanMetadata(), packageChannel: '_', packageUsername: '_' },
        },
      });

      expect(findConanMetadata().exists()).toBe(false);
    });
  });
});
