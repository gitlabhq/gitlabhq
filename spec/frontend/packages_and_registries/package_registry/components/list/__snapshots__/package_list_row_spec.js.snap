// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`packages_list_row renders 1`] = `
<div
  class="gl-border-b-1 gl-border-b-default gl-border-b-solid gl-border-t-1 gl-border-t-solid gl-border-t-transparent gl-flex gl-flex-col"
  data-testid="package-row"
>
  <div
    class="gl-flex gl-py-3"
  >
    <div
      class="gl-flex gl-justify-start gl-pt-2 gl-w-7"
    >
      <gl-form-checkbox-stub
        class="gl-m-0"
        value="true"
      />
    </div>
    <div
      class="@sm/panel:gl-flex-row gl-flex gl-flex-col gl-grow gl-items-stretch gl-justify-between"
    >
      <div
        class="@sm/panel:gl-mb-0 gl-flex gl-flex-col gl-grow gl-mb-3 gl-min-w-0"
      >
        <div
          class="@sm/panel:gl-flex gl-font-semibold gl-items-center gl-min-h-6 gl-min-w-0 gl-text-default"
        >
          <div
            class="gl-flex gl-gap-3 gl-items-center gl-min-w-0 gl-mr-5"
            data-testid="package-name"
          >
            <div
              class="gl-gap-2 sm:gl-flex"
            >
              <a
                class="gl-break-all gl-min-w-0 gl-text-default"
                data-testid="details-link"
              >
                @gitlab-org/package-15
              </a>
            </div>
          </div>
        </div>
        <div
          class="gl-flex gl-gap-3 gl-grow gl-items-center gl-min-h-6 gl-min-w-0 gl-text-sm gl-text-subtle"
        >
          <div
            class="gl-flex gl-items-center"
            data-testid="left-secondary-infos"
          >
            <gl-truncate-stub
              class="gl-max-w-15 md:gl-max-w-26"
              position="end"
              text="1.0.0"
              withtooltip="true"
            />
            <span
              class="gl-ml-2"
              data-testid="package-type"
            >
              Â· npm
            </span>
          </div>
        </div>
      </div>
      <div
        class="@sm/panel:gl-items-end gl-flex gl-flex-col gl-justify-center gl-shrink-0 gl-text-subtle"
      >
        <div
          class="@sm/panel:gl-text-default gl-flex gl-items-center gl-min-h-6"
        >
          <publish-method-stub />
        </div>
        <div
          class="gl-flex gl-items-center gl-min-h-6 gl-text-sm"
        >
          <publish-message-stub
            author=""
            projectname=""
            projecturl=""
            publishdate="2020-05-17T14:23:32Z"
          />
        </div>
      </div>
    </div>
    <div
      class="@sm/panel:gl-flex gl-items-center gl-pl-3"
    >
      <gl-disclosure-dropdown-stub
        autoclose="true"
        category="tertiary"
        data-testid="delete-dropdown"
        icon="ellipsis_v"
        items=""
        nocaret="true"
        placement="bottom-start"
        positioningstrategy="absolute"
        size="medium"
        textsronly="true"
        toggleid="dropdown-toggle-btn-2"
        toggletext="More actions"
        variant="default"
      >
        <gl-disclosure-dropdown-item-stub
          data-testid="action-delete"
          variant="danger"
        >
          Delete package
        </gl-disclosure-dropdown-item-stub>
      </gl-disclosure-dropdown-stub>
    </div>
  </div>
</div>
`;
